### ISSUE TITLE
seemless support for covalent ligands/protein mutations

### ISSUE THREAD

**Issue by jovhuffe** [openmm/openmm#3396]
Greetings

Is there an efficient way to use OpenMM to model covalent ligands and protein residue mutation? There doesn't seem any obvious way to do this without actually modifying the forcefield directly and/or using the generators indirectly. It's not as seemless as using them for non-covalent ligands.

Right now, the **workaround** could be to first rename the mutated protein residue topology, generate template for unmatched residues and somehow find the correct types by using template generators on a separate, equivalent non-covalent ligand and copy it onto the covalent counterpart because the generators do not seem to support covalent ligands directly.

happy holidays

**Comment by jchodera** [comment]
There's no simple, robust path for this at the moment, but there are a few efforts that are aiming to streamline this process:
* The [Open Force Field Initiative](http://openforcefield.org) is targeting this capability on its [2022 Science Roadmap](https://openforcefield.atlassian.net/wiki/spaces/PS/pages/423854401/Open+Force+Field+Scientific+Roadmap+2020?moved=true)
* Experimental tools like [espaloma](https://www.choderalab.org/publications/2020/10/6/end-to-end-differentiable-molecular-mechanics-force-field-construction) can easily handle arbitrary molecules (including covalent ligands) via graph convolutional machine learning models used to assign parameters, though I  will stress this is still highly experimental

**Comment by peastman** [comment]
A similar question came up recently in #3353.  A suggestion was made there to use AmberTools to build the model.

**Comment by jovhuffe** [comment]
Greetings, and thanks for the extremely quick responses. And happy to hear there's progress on that end.

I went ahead and tried the awesome **espaloma** package first.

> It works for small systems (3 residues including non-standard one) and seems to use the gpu for inference of the actual neural network. So this should mean it and my scripts are working properly.
> 
> However, for larger systems comprising of one protein (no solvent) it not only explodes the RAM usage when calling `espaloma.Graph(...)` but also takes extremely long to deploy by calling `espaloma.graphs.deploy.openmm_system_from_graph()`
> 
> Would this be because of not using the OpenEye toolkit?
> 
> I tried speeding things up by using numba  to no avail.
> Perhaps i'm missing something..?
> 
> **As I hoped to automate simulation for thousands of covalent ligands,** these build times are not feasible.
> I've read the paper as well but can't seem to find out what methods were used to speed things up in its writing.

using AmberTools:

> I wasn't too familiar with using AmberTools and related packages. But after closer inspection, it seems rather convoluted to automate parameteriziation as well, let alone within python.


tldr:
Espaloma works for small systems but takes too long for larger systems.

**Comment by jchodera** [comment]
@jovhuffe : Can you migrate this issue to the espaloma repo issue tracker?
https://github.com/choderalab/espaloma/issues

We'll be able to help you out with espaloma support there! If you have an example that we can help debug you can upload as a ZIP file, that would help us get you going! I suspect the issue is that your code was trying to use the AmberTools/OpenEye toolkit charges (which are infeasible for large systems) rather than espaloma charges, but it's possible there is something else going on.
